Doc/library/functools.rst

```diff
@@ -512,6 +518,192 @@ The :mod:`functools` module defines the following functions:
+
+   In case just an immediate sorting of the nodes in the graph is required and
+   no parallelism is involved, the convenience method :meth:`TopologicalSorter.static_order`
+   can be used directly. For example, this method can be used to implement a simple
+   version of the C3 linearization algorithm used by Python to calculate the Method
+   Resolution Order (MRO) of a derived class:
+
+   .. doctest::
+
+       >>> class A: pass
+       >>> class B(A): pass
+       >>> class C(A): pass
+       >>> class D(B, C): pass
+
+       >>> D.__mro__
+       (<class 'D'>, <class 'B'>, <class 'C'>, <class 'A'>, <class 'object'>)
+
+       >>> graph = {D: {B, C}, C: {A}, B: {A}, A:{object}}
+       >>> ts = TopologicalSorter(graph)
+       >>> topological_order = tuple(ts.static_order())
+       >>> tuple(reversed(topological_order))
+       (<class 'D'>, <class 'B'>, <class 'C'>, <class 'A'>, <class 'object'>)
+
+   The class is designed to easily support parallel processing of the nodes as they
+   become ready. For instance::
+
+       topological_sorter = TopologicalSorter()
+
+       # Add nodes to 'topological_sorter'...
+
+       topological_sorter.prepare()
+       while topological_sorter.is_active():
+           for node in topological_sorter.get_ready():
+               # Worker threads or processes take nodes to work on off the
+               # 'task_queue' queue.
+               task_queue.put(node)
+
+           # When the work for a node is done, workers put the node in
+           # 'finalized_tasks_queue' so we can get more nodes to work on.
+           # The definition of 'is_active()' guarantees that, at this point, at
+           # least one node has been placed on 'task_queue' that hasn't yet
+           # been passed to 'done()', so this blocking 'get()' must (eventually)
+           # succeed.  After calling 'done()', we loop back to call 'get_ready()'
+           # again, so put newly freed nodes on 'task_queue' as soon as
+           # logically possible.
+           node = finalized_tasks_queue.get()
+           topological_sorter.done(node)
+
+   .. method:: add(node, *predecessors)
+
+      Add a new node and its predecessors to the graph. Both the *node* and
+      all elements in *predecessors* must be hashable.
+
+      If called multiple times with the same node argument, the set of dependencies
+      will be the union of all dependencies passed in.
+
+      It is possible to add a node with no dependencies (*predecessors* is not
+      provided) or to provide a dependency twice. If a node that has not been
+      provided before is included among *predecessors* it will be automatically added
+      to the graph with no predecessors of its own.
+
+      Raises :exc:`ValueError` if called after :meth:`~TopologicalSorter.prepare`.
+
+   .. method:: prepare()
+
+      Mark the graph as finished and check for cycles in the graph. If any cycle is
+      detected, :exc:`CycleError` will be raised, but
```

Comment:
Unless I'm missing it, we never documented `CycleError`.  If so, salient points:

* It's a subclass of `ValueError`.
* If multiple cycles exist, it will only return one of them, and _which_ one is not defined.
* The cycle it detected can be accessed via accessing `.args[1]` of the exception instance.
   * And is a list of nodes, such that each node is, in the graph, an immediate predecessor of the next node in the list.
   * The first and last nodes of the list are the same, to make it clear that it's cyclic.